anaconda_packaging_utils/api/jira_api.py

```diff
@@ -0,0 +1,80 @@
+"""
+File:           jira_api.py
+Description:    Wrapper library that provides tools for using the Python JIRA API. The primary purpose of this wrapper
+                is to simplify and standardize the authentication process.
+
+                JIRA API links:
+                  - Source: https://github.com/pycontribs/jira
+                  - Docs: https://jira.readthedocs.io/index.html#
+"""
+import logging
+from typing import Callable, Final
+
+from jira.client import JIRA
+
+from anaconda_packaging_utils.storage.config_data import ConfigData
+
+# Logging object for this module
+log = logging.getLogger(__name__)
+
+# Where our JIRA boards are hosted
+_JIRA_HOST_URL: Final[str] = "https://anaconda.atlassian.net/"
+
+
+class ApiException(Exception):
+    """
+    Generic exception indicating an unrecoverable failure of this API.
+
+    This exception is meant to condense many possible failures into one generic error. The thinking is, if the calling
+    code runs into any API failure, there isn't much that can be done. So it is easier for the caller to handle one
+    exception than many exception types.
+    """
+
+    def __init__(self, message: str):
+        """
+        Constructs an API exception
+        :param message: String description of the issue encountered.
+        """
+        super().__init__(message if len(message) else "An unknown API issue was encountered.")
+
+
+class JiraApi:
+    """
+    Singleton wrapper to the Python Jira project. This "ensures" that we only construct and authenticate the underlying
+    Jira object once.
+    """
+
+    # The Jira API is wrapped in a list as a cheesy way to work around an initialization problem. Defaulting to `None`
+    # or using an `Optional` causes the static analyzer to freak out on every use of the `_jira`, even if the static
+    # variable has to have been initialized by instance-method call time.
+    _jira: list[JIRA] = []
+
+    def __init__(self) -> None:
+        """
+        Constructs a JiraApi instance
+        :raises ApiException: If there was a failure to authenticate.
+        """
+        if len(JiraApi._jira) == 0:
+            data_store: Final[ConfigData] = ConfigData()
+            try:
+                JiraApi._jira.append(
+                    JIRA(
+                        _JIRA_HOST_URL,
+                        basic_auth=(data_store["user_info.email"], data_store["token.jira"]),
+                    )
+                )
+            except Exception as e:
+                raise ApiException("Failed to auth or connect to JIRA") from e
+
+    def access_jira(self, callback: Callable[[JIRA], None]) -> None:
```

Comment:
I'm not sure I like this pattern. This is adding overhead compared to just giving full access to the client (the JIRA instance). As it is now, a caller would have to create a function just to call the Jira API or wrap the call into a lambda function. That's not very user friendly. And to be frank I'm also not sure how valuable it is to wrap these APIs anyway.

Reply:
This is why I added you to the PR, I had a feeling this might be a little too extreme and paranoid.

I think the wrappers serve a few goals:
1) They handle the auth process using a standardized and secure mechanism. If we find faults with the auth mechanism, we can control them all in one place without having to secure it in every project that invents their own auth management scheme.
2) They add common utilities that others may find helpful. I don't know what would be useful for a utility for the JIRA wrapper right now, but the GitHub one contains a few utilities to pull recipe files and parse them to our Percy recipe parser objects.
3) I attempt to distill down exceptions into a common and generic APIException per project. In sort of a "Trust but verify" mentality, I'm trying to prevent crashes caused by issues in 3rd party libraries. I'm sure I can't catch everything with this design, but I want to try to prevent breakages with uncaught exceptions.